### PR TITLE
perf: Skip the exact total count query on the last page of a search

### PR DESCRIPTION
**Pull request by John98Zakaria**
### 1. Why is this change necessary?

With `TOTAL_COUNT_MODE_EXACT`, `EntitySearcher` always runs the ids query and then a **second** `SELECT COUNT(*)` execution to obtain the total. When the requested page is the **last (partial) page**, that second query is redundant — the rows just fetched already determine the exact total.

### 2. What does this change do, exactly?

In `EntitySearcher::getTotalCount`, when the fetched page is not full (fewer rows than the limit, or no limit at all), the exact total is derived directly as `offset + rows` and the wrapped `COUNT(*)` query is skipped. Full pages (a further page may exist) and empty pages past the end (the total is unknown, only bounded by the offset) still fall back to the wrapped `COUNT(*)`.

The returned total and result set are byte-identical; this only removes a redundant query. It is unconditional, adds no work to any query (no window functions, no extra columns), and touches a single method.

### 3. Measured impact

30 000 products, `TOTAL_COUNT_MODE_EXACT`, an intentionally expensive filter (six matching `LIKE`s so the skipped `COUNT(*)` is measurable), MySQL 8.0, median of 9 runs across `EntityRepository::search()`:

| request | before | after |
|---|--:|--:|
| last (partial) page (`offset 29990`, `limit 25`) | 187 ms | **131 ms** (−30%) |
| full page mid-list, control (`offset 0`, `limit 25`) | 63 ms | 66 ms (unchanged) |

The win scales with how expensive the search's `COUNT(*)` is; it is negligible for a cheap indexed listing and removes a full second execution for an expensive filtered search. The control confirms the change never affects searches where the short-circuit does not apply.

### 4. How to test it

`EntitySearcherTest::testExactTotalCountShortCircuitsOnTheLastPage` (data provider) asserts the exact total equals the real number of matching records across every pagination boundary: partial last page, single partial page, no limit, full pages, and an empty page past the end.

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have documented my changes in `RELEASE_INFO-6.7.md` and `UPGRADE-6.8.md`
- [x] This change has comments for non-obvious lines of code
- [x] I have read the contribution requirements and fulfill them

### CODE REVIEW

File: RELEASE_INFO-6.7.md [hunk 1]
```diff
@@ -6,6 +6,10 @@
 
 ## Core
 
+### Skip the exact total count query on the last page of a search
```
Comment (mitelg) [inline RELEASE_INFO-6.7.md]: the change is not directly 3rd party developer relevant, so this can be reverted

File: UPGRADE-6.8.md [hunk 2]
```diff
@@ -253,6 +253,10 @@ Previously, these routes could return unrelated records or fail because the unde
 
 <details>
 
+## Exact total count skips the COUNT query on the last page
```
Comment (mitelg) [inline UPGRADE-6.8.md]: as nothing changes for the major update in 6.8, an entry in the upgrade guide is not necessary -> revert

**Review by mitelg — CHANGES_REQUESTED** [review]

**Comment by timtheisinger** [comment]
@mitelg Hi, John is on vacation. I removed the text changes

File: tests/integration/Core/Framework/DataAbstractionLayer/Search/EntitySearcherTest.php [hunk 3]
```diff
@@ -331,6 +332,63 @@ public function testSortingAndTotalCountWithManyAssociation(): void
         static::assertCount(6, $result->getEntities());
     }
 
+    /**
+     * @param array{offset: int, limit: int|null, expectedEntities: int} $pagination
+     */
+    #[DataProvider('lastPagePaginationProvider')]
+    public function testExactTotalCountShortCircuitsOnTheLastPage(array $pagination): void
```
Comment (mitelg) [inline tests/integration/Core/Framework/DataAbstractionLayer/Search/EntitySearcherTest.php:365]: the tests are not ensuring that the DB query is not executed. it ensures correct results, but they were correct before as well. 

would you mind adding another (unit) test?

**Review by mitelg — CHANGES_REQUESTED** [review]

File: tests/integration/Core/Framework/DataAbstractionLayer/Search/EntitySearcherTest.php (hunk 3 above)
Comment (John98Zakaria) [reply tests/integration/Core/Framework/DataAbstractionLayer/Search/EntitySearcherTest.php:365]: Sure I will add one more test

File: tests/unit/Core/Framework/DataAbstractionLayer/Dbal/EntitySearcherTest.php [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+            static::createStub(ValidatorInterface::class),
+            static::createStub(EntityWriteGateway::class)
+        );
+    }
+
+    #[DataProvider('lastPageProvider')]
+    public function testExactTotalCountSkipsTheCountQueryOnTheLastPage(?int $offset, ?int $limit, int $rows, int $expectedTotal): void
+    {
+        $connection = $this->createMock(Connection::class);
+        $connection->method('getDatabasePlatform')->willReturn(new MySQLPlatform());
```
Comment (mitelg) [inline tests/unit/Core/Framework/DataAbstractionLayer/Dbal/EntitySearcherTest.php]:
```suggestion
```
this looks unrelated

File: tests/unit/Core/Framework/DataAbstractionLayer/Dbal/EntitySearcherTest.php [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        // Without an offset an empty page proves that nothing matches at all.
+        yield 'empty first page' => [0, 25, 0, 0];
+        yield 'empty unpaginated criteria' => [null, null, 0, 0];
+    }
+
+    #[DataProvider('notLastPageProvider')]
+    public function testExactTotalCountFallsBackToTheCountQuery(?int $offset, int $limit, int $rows): void
+    {
+        $connection = $this->createMock(Connection::class);
+        $connection->method('getDatabasePlatform')->willReturn(new MySQLPlatform());
```
Comment (mitelg) [inline tests/unit/Core/Framework/DataAbstractionLayer/Dbal/EntitySearcherTest.php]:
```suggestion
```

same here. tested it locally, and the tests will work without that

**Review by mitelg — APPROVED** [review]

File: tests/integration/Core/Framework/DataAbstractionLayer/Search/EntitySearcherTest.php [hunk 6]
```diff
@@ -357,6 +358,63 @@ public function testSortingAndTotalCountWithManyAssociation(): void
...
+    {
+        $context = Context::createDefaultContext();
+
+        $totalMatching = 6;
```
Comment (nfortier-shopware) [inline tests/integration/Core/Framework/DataAbstractionLayer/Search/EntitySearcherTest.php]: If we need to test strictly pagination in integration, it is better to do it with the minimum products like 3.

something like that (if possible)

```php
public static function lastPagePaginationProvider(): iterable
{
    // Partial last page (Offset 2 + 1 item remaining = 3 total)
    yield 'partial last page' => [['offset' => 2, 'limit' => 2, 'expectedEntities' => 1]];

    // Single partial page (Offset 0 + 3 items = 3 total, which is < limit 25)
    yield 'single partial page' => [['offset' => 0, 'limit' => 25, 'expectedEntities' => 3]];

    // No limit at all
    yield 'no limit' => [['offset' => 0, 'limit' => null, 'expectedEntities' => 3]];

    // Full page with more remaining (Offset 0 + limit 1 = 1 item returned, 2 remain)
    yield 'first of several full pages' => [['offset' => 0, 'limit' => 1, 'expectedEntities' => 1]];

    // Exactly full last page (Offset 2 + limit 1 = 1 item returned, 0 remain)
    yield 'exactly full last page' => [['offset' => 2, 'limit' => 1, 'expectedEntities' => 1]];

    // Empty page past the end (Offset 5 is greater than 3 total)
    yield 'empty page past the end' => [['offset' => 5, 'limit' => 1, 'expectedEntities' => 0]];
}
````

File: tests/unit/Core/Framework/DataAbstractionLayer/Dbal/EntitySearcherTest.php (hunk 4 above)
Comment (nfortier-shopware) [reply tests/unit/Core/Framework/DataAbstractionLayer/Dbal/EntitySearcherTest.php]:
```php 
$connection = $this->createStub(Connection::class);
$connection->method('executeQuery')->willReturn($this->createIdResult($rows));
```

Unless it is critical to verify that the query is executed only one time

Comment (mitelg) [reply tests/unit/Core/Framework/DataAbstractionLayer/Dbal/EntitySearcherTest.php]: yes, that is the whole point of the unit test! :sweat_smile: please leave the assertion on the call times

Comment (nfortier-shopware) [reply tests/unit/Core/Framework/DataAbstractionLayer/Dbal/EntitySearcherTest.php]: it was not obvious but I had suspicion (more because of the other test with exactly 2)

**Review by nfortier-shopware — APPROVED** [review]

File: tests/unit/Core/Framework/DataAbstractionLayer/Dbal/EntitySearcherTest.php (hunk 4 above)
Comment (shyim) [reply tests/unit/Core/Framework/DataAbstractionLayer/Dbal/EntitySearcherTest.php]: Applied: removed the unused `getDatabasePlatform` stubs and kept the `executeQuery` call-count assertions.

File: tests/unit/Core/Framework/DataAbstractionLayer/Dbal/EntitySearcherTest.php (hunk 5 above)
Comment (shyim) [reply tests/unit/Core/Framework/DataAbstractionLayer/Dbal/EntitySearcherTest.php]: Applied here as well.

File: tests/integration/Core/Framework/DataAbstractionLayer/Search/EntitySearcherTest.php (hunk 6 above)
Comment (shyim) [reply tests/integration/Core/Framework/DataAbstractionLayer/Search/EntitySearcherTest.php]: Applied: the integration fixture now uses three products and the pagination cases you suggested.

**Review by shyim — APPROVED** [review]

**Comment by shyim** [comment]
Thanks for the patch!
